### PR TITLE
JobsDSL: better graph building algorithm for flows

### PR DESCRIPTION
**Pull request by aclement**
- improved text to graph creation for flows
- improved DSL creation from graph
- Fixed tokenizer problem when whitespace not supplied

### CODE REVIEW

**Comment by cppwfs** [comment]
In Graph.java we can merge the the if statements on line 190 and 194 into a single if statement.

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dsl/Graph.java [hunk 1]
```diff
@@ -189,10 +198,11 @@ private void followLinks(StringBuilder graphText, List<Link> toFollow, Node node
 				unvisitedNodes.remove(endOfSplit);
 				if (!endOfSplit.isSync()) {
 					// If not a sync node, include it in the output text
-					graphText.append(" || ");
+					graphText.append(" || "); 
 					printNode(graphText, endOfSplit, unvisitedNodes);
 					List<Link> transitionalLinks = findLinksFrom(endOfSplit, false);
-					printTransitions(graphText, unfollowedLinks, transitionalLinks);
+					// null final param here probably not correct
+					printTransitions(graphText, unvisitedNodes, unfollowedLinks, transitionalLinks, null);
```
Comment (cppwfs) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dsl/Graph.java]: Null is not necessarily evil, but FYI: Sonar may tag us on changing the state on objects that are passed in via parameters into methods.

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dsl/Graph.java [hunk 2]
```diff
@@ -200,23 +210,43 @@ else if (toFollow.size() == 1) { // FLOW
...
 					}
+					followLink(graphText, linkToFollow, nodeToTerminateFollow, unvisitedNodes, unfollowedLinks);
 				}
 				break;
 			}
 		}
 	}
 
+	private String toString(Link link) {
+		StringBuilder s = new StringBuilder();
```
Comment (cppwfs) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dsl/Graph.java]: Not sure we need to use StringBuilder in this method to create our string in this method.

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dsl/Graph.java [hunk 3]
```diff
@@ -200,23 +210,43 @@ else if (toFollow.size() == 1) { // FLOW
...
+				s.append("::");
+			}
+			s.append(toString(links.get(i)));
+		}
+		s.append("]");
+		return s.toString();
+	}
+
 	private Node findEndOfSplit(List<Link> toFollow) {
+		//DEBUG: System.out.println("Finding end of split, incoming links " + toString(toFollow));
```
Comment (cppwfs) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dsl/Graph.java]: we can remove this comment

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dsl/JobSpecification.java [hunk 4]
```diff
@@ -929,9 +945,32 @@ private int findNode(String targetJob) {
...
+				// Not already in this flow, create a new one
+				nextId = id++;
+				node = new Node(Integer.toString(nextId), jr.getName(), null, properties);
+				nodes.add(node);
+			}
 			createdNodes.put(jr.getName(), node);
+
+			// Should nodes created in the mainline visit be inserted into the shared set?
+			// (In addition to those created for transitions). Do we need it to support
+			// back references (loops?)
```
Comment (cppwfs) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dsl/JobSpecification.java]: good question.  Let's ask Michael.  My inclination is no...

Comment (aclement) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dsl/JobSpecification.java]: I agree no for right now - left this in more as a reminder to say when someone asks for loops, might need to do something here.

**Comment by cppwfs** [comment]
LGTM, Squashed, Rebased, Merged

**Pull request closed without merging**